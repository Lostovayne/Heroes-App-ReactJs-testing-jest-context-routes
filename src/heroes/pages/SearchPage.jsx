import { useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import queryString from 'query-string';
import { HeroCard } from '../components/HeroCard';
import { useLocation } from 'react-router-dom';
import { getHeroByName } from '../helpers';

export const SearchPage = () => {
    const navigate = useNavigate();
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const heroes = getHeroByName(q);

    const { searchText, onInputChange, onResetForm } = useForm({
        searchText: q,
    });

    const onSearchSubmit = (event) => {
        event.preventDefault();
        navigate(`?q=${searchText}`);
        onResetForm();
    };

    return (
        <>
            <h1>Search</h1>
            <hr />
            <div className='row'>
                <div className='col-5'>
                    <h4>Searching</h4>
                    <hr />
                    <form onSubmit={onSearchSubmit}>
                        <input
                            type='text'
                            placeholder='Search a hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={onInputChange}
                        />
                        <button className='btn btn-outline-primary' type='submit'>
                            Search
                        </button>
                    </form>
                </div>

                <div className='col-7'>
                    <h4>Results</h4>
                    <hr />

                    {q === '' && <div className='alert alert-primary'>Search a hero</div>}
                    {heroes.length === 0 && q !== '' && (
                        <div className='alert alert-danger'>No hero with {q}</div>
                    )}

                    {heroes.map((hero) => (
                        <HeroCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </>
    );
};
