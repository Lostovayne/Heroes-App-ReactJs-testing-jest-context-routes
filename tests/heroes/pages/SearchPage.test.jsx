import {SearchPage} from './../../../src/heroes/pages/SearchPage';
import {MemoryRouter} from 'react-router-dom';
import {render, screen, fireEvent} from '@testing-library/react';
import {useNavigate} from 'react-router-dom';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe(' Pruebas en Search Page', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('Debe de mostrarse correctamente', () => {
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
        // screen.debug();
    });

    test('Buscar un heroe', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=spider']}>
                <SearchPage />
            </MemoryRouter>
        );
        // el input de busqueda debe de existir con el valor spider
        expect(screen.getByDisplayValue('spider')).toBeTruthy();
        // console.log(screen.getByDisplayValue('spider').value);
        const Input = screen.getByRole('textbox');
        expect(Input.value).toBe('spider');
        const Img = screen.getByRole('img');
        expect(Img.src).toContain('spider');
        //evaluar el texto search a hero no deberia aparecer
        expect(screen.queryByText('search a hero')).toBeNull();
        const divHeroBody = screen.getByLabelText('HeroBody');
        expect(divHeroBody.className).toContain('card-body');
    });

    test('Debe de mostrar un error si no se encuentra el heroe batman123', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );
        // screen.debug();
        expect(screen.getByText('No hero with batman123')).toBeTruthy();
        expect(screen.queryByText('search a hero')).toBeNull();
        const HeroAlert = screen.getByLabelText('HeroAlert');
        expect(HeroAlert.className).toContain('alert-danger');
        expect(HeroAlert.textContent).toContain('No hero with batman123');
    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        // screen.debug();

        const InputHero = screen.getByRole('textbox');
        fireEvent.change(InputHero, {
            target: {
                name: 'searchText',
                value: 'superman',
            },
        });

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman');
    });
});
