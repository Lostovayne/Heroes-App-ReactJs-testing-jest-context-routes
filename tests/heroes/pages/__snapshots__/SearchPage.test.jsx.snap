// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Pruebas en Search Page Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
